atomicapp/requirements.py

```diff
@@ -0,0 +1,75 @@
+import logging
+
+from atomicapp.constants import REQUIREMENT_FUNCTIONS
+from atomicapp.plugin import Plugin
+
+plugin = Plugin()
+plugin.load_plugins()
+
+logger = logging.getLogger(__name__)
+
+
+class Requirements:
+
+    """
+    Requirements will search what is currently being used under
+    the requirements section of Nulecule and deploy it in according
+    to the graph variables as well as whether or not said requirement
+    exists within the provider.
+
+    The REQUIREMENTS_FUNCTIONS dictionary maps all current requirement
+    names to the function names for each provider.
+
+    For example, the persistentVolume requirement in Nulecule is mapped as
+    the persistent_storage function within each provider.
+
+    Requirements tries to be as modular as possible.
+    """
+
+    def __init__(self, config, basepath, graph, provider, dryrun):
+        self.config = config
+        self.basepath = basepath
+        self.graph = graph
+        self.dryrun = dryrun
+
+        # We initialize the provider in order to gather essential information such as kubectl config
+        self.provider = plugin.getProvider(provider)(config, basepath, dryrun)
+        self.provider.init()
+
+    def run(self):
+        self._exec("run")
+
+    def stop(self):
+        self._exec("stop")
+
+    def uninstall(self):
+        self._exec("uninstall")
+
+    # Find if the requirement does not exist within REQUIREMENT_FUNCTIONS
+    def _find_requirement_name(self, key):
+        logging.debug("Checking if %s matches any of %s" % (key, REQUIREMENT_FUNCTIONS))
+        if key in REQUIREMENT_FUNCTIONS.keys():
+            return REQUIREMENT_FUNCTIONS[key]
+        raise RequirementFailedException("Requirement %s does not exist." % key)
+
+    # We loop through the given requirements graph and
+    # execute each passed requirement
+    def _exec(self, action):
+        for req in self.graph:
+            key_name = req.keys()[0]
+            requirement_name = self._find_requirement_name(key_name)
+
+            # Check to see if the function exists in the provider, if it does not: fail
+            try:
+                requirement_function = getattr(self.provider, requirement_name)
+            except AttributeError:
+                raise RequirementFailedException(
+                    "Requirement %s does not exist within %s." % (requirement_name, self.provider))
+
+            # Run the requirement function
+            requirement_function(req[key_name], action)
+
+
+class RequirementFailedException(Exception):
+
```

Comment:
do we not need to put "pass" here like we do for the other exceptions we have defined?